lib/spack/spack/util/spack_yaml.py

```diff
@@ -284,37 +319,133 @@ def write_line_break(self):
+    ANNOTATED_SPACK_CONFIG_FILE = enum.auto()
+
+
+class ConfigYAML:
+    """Handles the loading and dumping of Spack's YAML files."""
+
+    def __init__(self, yaml_type: YAMLType) -> None:
+        self.yaml = ruamel.yaml.YAML(typ="rt", pure=True)
+        if yaml_type == YAMLType.GENERIC_YAML:
+            self.yaml.Representer = SafeRepresenter
+        elif yaml_type == YAMLType.ANNOTATED_SPACK_CONFIG_FILE:
+            self.yaml.Representer = LineAnnotationRepresenter
+            self.yaml.Emitter = LineAnnotationEmitter
+            self.yaml.Constructor = OrderedLineConstructor
         else:
-            _annotations.append("")
+            self.yaml.Representer = OrderedLineRepresenter
+            self.yaml.Constructor = OrderedLineConstructor
+
+    def load(self, stream: IO):
+        """Loads the YAML data from a stream and returns it.
+
+        Args:
+            stream: stream to load from.
+
+        Raises:
+            SpackYAMLError: if anything goes wrong while loading
+        """
+        try:
+            return self.yaml.load(stream)
+
+        except error.MarkedYAMLError as e:
+            msg = "error parsing YAML"
+            error_mark = e.context_mark if e.context_mark else e.problem_mark
+            if error_mark:
+                line, column = error_mark.line, error_mark.column
+                msg += f": near {error_mark.name}, {str(line)}, {str(column)}"
+            else:
+                msg += f": {stream.name}"
+            msg += f": {e.problem}"
+            raise SpackYAMLError(msg, e) from e
+
+        except Exception as e:
+            msg = "cannot load Spack YAML configuration"
+            raise SpackYAMLError(msg, e) from e
+
+    def dump(self, data, stream: Optional[IO] = None, *, transform=None) -> None:
+        """Dumps the YAML data to a stream.
+
+        Args:
+            data: data to be dumped
+            stream: stream to dump the data into.
+
+        Raises:
+            SpackYAMLError: if anything goes wrong while dumping
+        """
+        try:
+            return self.yaml.dump(data, stream=stream, transform=transform)
+        except Exception as e:
+            msg = "cannot dump Spack YAML configuration"
+            raise SpackYAMLError(msg, str(e)) from e
+
+    def as_string(self, data) -> str:
+        """Returns a string representing the YAML data passed as input."""
+        result = io.StringIO()
+        self.dump(data, stream=result)
+        return result.getvalue()
+
+
+def deepcopy(data):
+    """Returns a deepcopy of the input YAML data."""
+    result = copy.deepcopy(data)
+
+    if isinstance(result, comments.CommentedMap):
+        # HACK to fully copy ruamel CommentedMap that doesn't provide copy
+        # method. Especially necessary for environments
+        setattr(
+            result,
+            comments.Comment.attrib,
+            getattr(data, comments.Comment.attrib, comments.Comment()),
+        )
```

Comment:
This could probably be cleaned up by using `extract_comment` and `set_comment` so there's less to do if the comment format changes later. Maybe something like

```suggestion
        comments = extract_comments(data)
        if comments:
            set_comments(result, data_comments=comments)
```

Reply:
Fixed in 9adbd9a